packages/utils/typescript/lib/generators/index.js

```diff
@@ -1,7 +1,124 @@
+const { saveDefinitionToFileSystem, createLogger, timer } = require('./utils');
+const generateContentTypesDefinitions = require('./content-types');
+const generateComponentsDefinitions = require('./components');
+
+const GENERATORS = {
+  contentTypes: generateContentTypesDefinitions,
+  components: generateComponentsDefinitions,
+};
+
+/**
+ * @typedef GenerateConfig
+ *
+ * @property {object} strapi
+ * @property {boolean} pwd
+ * @property {object} [artefacts]
+ * @property {boolean} [artefacts.contentTypes]
+ * @property {boolean} [artefacts.components]
+ * @property {boolean} [artefacts.services]
+ * @property {boolean} [artefacts.controllers]
+ * @property {boolean} [artefacts.policies]
+ * @property {boolean} [artefacts.middlewares]
+ * @property {object} [logger]
+ * @property {boolean} [logger.silent]
+ * @property {boolean} [logger.debug]
+ * @property {boolean} [logger.verbose]
+ */
+
+/**
+ * Generate types definitions based on the given configuration
+ *
+ * @param {GenerateConfig} [config]
+ */
+const generate = async (config = {}) => {
+  const { pwd, rootDir = TYPES_ROOT_DIR, strapi, artefacts = {}, logger: loggerConfig } = config;
+  const reports = {};
+  const logger = createLogger(loggerConfig);
+  const psTimer = timer().start();
+
+  const registryPwd = path.join(pwd, rootDir, REGISTRIES_OUT_DIR);
+  const generatorConfig = { strapi, pwd: registryPwd, logger };
+
+  const returnWithMessage = () => {
+    const nbWarnings = chalk.yellow(maybePlural('warning', logger.warnings));
+    const nbErrors = chalk.red(maybePlural('error', logger.errors));
+
+    const status = logger.errors > 0 ? chalk.red('errored') : chalk.green('completed successfully');
+
+    psTimer.end();
+
+    logger.info(`The task ${status} with ${nbWarnings} and ${nbErrors} in ${psTimer.duration}s.`);
+
+    return reports;
+  };
+
+  const enabledArtefacts = Object.keys(artefacts).filter((p) => artefacts[p] === true);
+
+  logger.info('Starting the type generation process');
+  logger.debug(`Enabled artefacts: ${enabledArtefacts.join(', ')}`);
+
+  for (const artefact of enabledArtefacts) {
+    const boldArtefact = chalk.bold(artefact); // used for log messages
+
+    logger.info(`Generating types for ${boldArtefact}`);
+
+    if (artefact in GENERATORS) {
+      const generator = GENERATORS[artefact];
+
+      try {
+        const artefactGenTimer = timer().start();
+
+        reports[artefact] = await generator(generatorConfig);
+
+        artefactGenTimer.end();
+
+        logger.debug(`Generated ${boldArtefact} in ${artefactGenTimer.duration}s`);
+      } catch (e) {
+        logger.error(
+          `Failed to generate types for ${boldArtefact}: ${e.message ?? e.toString()}. Exiting`
+        );
+        return returnWithMessage();
+      }
+    } else {
+      logger.warn(`The types generator for ${boldArtefact} is not implemented, skipping`);
+    }
+  }
+
+  for (const artefact of Object.keys(reports)) {
```

Comment:
Here's some to discuss – US spelling is `artifact` imo we use US spelling in coding e.g. `color` so I think it makes sense to keep that – wdyt?

Reply:
I don't think we have an official policy on which English dialect to use (a quick search for "ised" vs "ized" will show hundreds of differences) so I'm fine with artefact until someone says otherwise.

Reply:
> a quick search for "ised" vs "ized" will show hundreds of differences 

gutted. But makes sense 👍🏼

Reply:
I mean, artefact did throw me a bit, I'd say it's even rarer to see in US english than -ised or colour. So switching to artifact would make my brain more comfortable but other than that I don't have a valid justification

Reply:
This one's funny since I tried to make some research to avoid this comment as I wasn't sure at all what was the best one to use, and apparently, "artifact" was the most common one.

However I just checked on Trends, and [it appears](https://trends.google.com/trends/explore?date=all&q=artifact,artefact&hl=en) that artifact is preferred :sweat_smile: 
I'll change the wording

Reply:
I didn't realise `artefact` was the UK spelling to be honest.